README.md
# T-Fold-SV
T-Fold Sequential Validation Technique, a safe replacement for K-Fold when using Financial Time Series Data.


CHANGELOG.rst

Changelog
=========

------------
21.June.2021
------------

- Added: ``ohlc_labeling`` for target variable formation.
- Modified: format and lables in `continuous_futures.py` and `synthetic.py`
- Modified: ``kld`` to shift data to have positive values, docstrings, method and sub-methods visibility.

------------
20.June.2021
------------

- Added: RandomWalk Time Series Generator (Basic version).
- Added: 10+ years of OHLC prices + Volume Continuous Futures files (2010 - 2021) (UsdMxn and EurUsd).
- Added: ``resample_data`` function for prices downsample based on timestamp.
- Added: ``folds_formation`` to produce 'quarter', 'semester', 'year', '80-20' Fold sizes.
- Added: ``kld`` to measure the divergence between two PDFs using Kullback-Liebler Divergence metric.

------------
09.June.2021
------------

- Added: File docstrings and general comments
- Added: Image files for README.rst


main.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- T-Fold-SV is Time Series Folds for Sequential Validation, the go to alternative for K-Fold-CV       -- #
# -- --------------------------------------------------------------------------------------------------- -- #
# -- Description: Python Implementation of the T-Fold Sequential Validation Method                       -- #
# -- main.py: python script with main method                                                             -- #
# -- Author: IFFranciscoME - if.francisco.me@gmail.com                                                   -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- Repository: https://github.com/IFFranciscoME/T-Fold-SV                                              -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

# -- Load libraries for script
import pandas as pd
import numpy as np

# -- Load other scripts
import data as dt
import functions as fn

# -- Read Future Continuous Prices -- #

# File name and route
file_name = 'MP_H1_2010_2021'
file_route = 'files/prices/' + file_name + '.txt'

# Read files and prepare format
data_futures = pd.read_csv(file_route, header=None,
                           names=['timestamp', 'open', 'high', 'low', 'close', 'volume'])

# Conversion from Usd per Mxn to Mxn per Usd
data_futures['timestamp'] = pd.to_datetime(data_futures['timestamp'])
data_futures['open'] = round(1/data_futures['open'], 5)
data_futures['high'] = round(1/data_futures['high'], 5)
data_futures['low'] = round(1/data_futures['low'], 5)
data_futures['close'] = round(1/data_futures['close'], 5)

# Swap high and low since the exchange rate was swapped from MXNUSD to USDMXN
low = data_futures['low'].copy()
high = data_futures['high'].copy()
data_futures['high'] = low
data_futures['low'] = high

# Clean memory for these dummy objects
high, low = None, None

# -- 0. Load Global Dataset (Downsample price data)
global_data = dt.resample_data(target_data=data_futures.copy(), target_freq='8H',
                               auto_names=True, col_names=None)

"""
Pre-Building

- an OHLCV timestamp-indexed dataset.
- Specify the fold_size parameter.
- Specification for labeling the target variable.
- Specification for feature engineering process.
"""

# ----- T-Fold-SV ----- # 
# --------------------- #
"""

- Information Leakage
- - Criteria specification (embargo and purge)

- Information Tensor
- - PDF with target variable in every Fold
- - KLD with all combinations of Folds
- - Information Threshold

- Sparsity assesment
- - Classify Information tensor as: Sparse, Weakly Sparse, Non-Sparse

- Learning Process
- - Define type of learning: Sparse, Weakly Sparse, Non-Sparse
- - Model definition (parameters, hyperparameters)
- - Conduct Optimization and Learning processes

- Performance assesment
- - Define performance metric (Context dependant)
- - Out-of-Sample generalization assesment
- - Out-of-Distribution generalization assesment

Type 2 requires:

- A timestamp-indexed dataset
- At least one target variable, and one or more Features.
- Specify the fold_size parameter.
- Specify Information Leakage Prevention Criteria.
"""

# ---------------------------------------------------------------------------------- T-FOLD-SV : TYPE 1 -- #  
# ---------------------------------------------------------------------------------- ------------------ -- #  

# -- 1. Folds Formation
folds_data = dt.folds_formation(global_data=global_data, fold_size='year')

# -- 2. Target and Features Engineering

# Target Variable: Time-based labeling
Targets = {}
for label in list(folds_data.keys()):
    Targets[label] = fn.ohlc_labeling(ohlc_data=folds_data[label], p_label='co')

# Feature Engineering
Features = {}
for label in list(folds_data.keys()):
    Features[label] = fn.ohlc_features(ohlc_data=folds_data[label], p_label='co')

# -- Tests for target variable formation and KLD metric for continuous variables
target_2010 = fn.ohlc_labeling(ohlc_data=folds_data['y_2010'], p_label='co')
target_2011 = fn.ohlc_labeling(ohlc_data=folds_data['y_2011'], p_label='co')
kld_metric = fn.kld(p_data=target_2010, q_data=target_2011, prob_dist='gamma')

# Feature Engineering: In-Fold + Heuristic

# -- 3. Information Assesment
# -- 4. Learning Assesment
# -- 5. Generalization Assesment

# ------------------------------- #


functions.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- T-Fold-SV is Time Series Folds for Sequential Validation, the go to alternative for K-Fold-CV       -- #
# -- --------------------------------------------------------------------------------------------------- -- #
# -- Description: Python Implementation of the T-Fold Sequential Validation Method                       -- #
# -- functions.py: Mathematical and Data Processing functions                                            -- #
# -- Author: IFFranciscoME - if.francisco.me@gmail.com                                                   -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- Repository: https://github.com/IFFranciscoME/T-Fold-SV                                              -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

# -- Load libraries for script
import numpy as np
import pandas as pd
import scipy.special as sps

# ---------------------------------------------------------------------------------------- OHLC LABELING -- #
# --------------------------------------------------------------------------------------------------------- #

def ohlc_labeling(ohlc_data, p_label='b_co'):
    """
    This methods offers some options to generate target variables according to the selected labeling process. The input, OHLC prices, is already time-based labelled, nevertheless, a lower granularity labeling process can be conducted. It is recomended though to use this function with higher frequency prices (by the minute or more frequent if possible).

    Normally, a numeric value like co = close - open will be utilized for a regression type of problem, whereas the binary classification problem can be formulated as the sign operation for co, having therefore a 1 if close > open and 0 otherwise. For numerical stability of some cost functions, it is better to have 1 and 0 instead of 1 and -1.

    Parameters
    ----------

    ohlc_data: DataFrame
        With at least 4 numeric columns: 'open', 'high', 'low', 'close'
    
    p_label: str
        An indication of the labelling function, must chose one of the following options:
            'co': close - open
            'b_co': binary version of 'co', i.e. sign[close - open], 1 if close > open, 0 otherwise

    Returns
    -------

        labels according to selected method, currently the results will be:
            - numeric when selecting 'co'
            - numeric and binary (1s and 0s) when selecting 'b_co'

    Example
    -------

    >>> ohlc_data = pd.DataFrame({'timestamp': ['2020-01-01 10:00:00', '2020-01-01 10:00:00'],
                                 'open': [1.25, 1.25], 'high': [1.35, 1.35],
                                 'low': [1.15, 1.15], 'close': [1.30, 1.30], 'vol': [123, 123]})
    >>> label = 'b_co'

    """

    # shallow copy of data
    f_data = ohlc_data.copy()
    # base calculation
    co = f_data['close'] - f_data['open']
    # return continuous variable
    if p_label == 'co':
        return co
    # return discrete-binary variable
    elif p_label == 'b_co':
        return pd.Series([1 if i > 0 else 0 for i in list(co)], index=f_data.index)
    
    # raise error
    else:
        raise ValueError("Accepted values for label are: 'co' or 'b_co' ")

# -------------------------------------------------------------------------- KULLBACK-LEIBLER DIVERGENCE -- #
# --------------------------------------------------------------------------------------------------------- #

def kld(p_data, q_data, prob_dist, pq_shift=True):
    """
    Computes the divergence between two empirical adjusted probability density functions.

    Parameters
    ----------

    p_data: np.array
        Data of the first process

    q_data: np.array
        Data of the first process
    
    prob_dist: str
        Probability distribution - Added: to fit to empirical data
        'gamma': Generalized gamma distribution
    
    pq_shift: bool
        True (Default): Shifts the data in order to have only positive values. This is done by adding, to all values, the absolute of the most negative value.
    
    Returns
    -------
    
    r_kld_gamma: numeric
            Kullback-Leibler Divergence Metric

    References
    ----------

    [1] Kullback, S., & Leibler, R. (1951). On Information and Sufficiency. The Annals of Mathematical  Statistics, 22(1), 79-86. Retrieved June 21, 2021, from http://www.jstor.org/stable/2236703

    Example
    -------

    >>> data_p = np.random.default_rng().gamma(2, 1, 100)
    >>> data_q = np.random.default_rng().gamma(1, 2, 100)
    >>> kld_metric = kld(p_data=data_p, q_data=data_q, prob_dist='gamma')

    """

    # Shift data to have only positive values
    if pq_shift:
        q_data = (q_data + abs(min(q_data)))/max(q_data)
        p_data = (p_data + abs(min(p_data)))/max(p_data)

    # -- with Gamma Distribution -- #
    # ----------------------------- #

    # For continuous variables
    if prob_dist == 'gamma':
        return _kld_gamma(p_data=p_data, q_data=q_data)

    # -- with Binomial Distribution -- #
    # -------------------------------- #

    # For discrete variables (Pending)

    # -- with Other Distribution -- #
    # ----------------------------- #

    else:
        return print('error')

# ----------------------------------------------------------------------- KLD with generalized gamma -- #

def _kld_gamma(p_data, q_data):
    """
    Computes the Kullback-Leibler divergence between two gamma PDFs
    
    Parameters
    ----------

    p_data: np.array
        Data of the first process

    q_data: np.array
        Data of the first process

    Returns
    -------

    r_kld_gamma: numeric
        Kullback-Leibler Divergence Quantity
    
    References
    ----------
    [1] Bauckhage, Christian. (2014). Computing the Kullback-Leibler Divergence between two Generalized Gamma Distributions. arXiv. 1401.6853. 
    
    """
    
    # -------------------------------------------------------------------------- Distribution Parameters -- #

    def _gamma_params(data, method='MoM'):
        """
        Computes the parameters of a gamma probability density function (pdf), according to the selected
        method.

        Parameters
        ----------

        data: np.array
            The data with which will be adjusted the pdf
        
        method: str
            Method to calculate the value of the parameters for the pdf
                'MoM': Method of Moments (Default)

        Returns
        -------

        r_params: dict
            {'alpha': gamma distribution paramerter, 'beta': gamma distribution parameter}
        
        """

        # -- Methods of Moments -- #
        if method == 'MoM':

            # first two moments
            mean = np.mean(data)
            variance = np.var(data)
            # sometimes refered in literature as k
            alpha = mean**2/variance
            # sometimes refered in literature as 1/theta
            beta = mean/variance
            # return the gamma distribution empirically adjusted parameters
            return alpha, beta
        
        # -- For errors or other unsupported methods
        else:
            raise ValueError("Currently, the supported methods are: 'MoM'")

    # alpha_1: Distribution 1: shape parameter, alpha_1 > 0
    # beta_1:  Distribution 1: rate or inverse scale distribution parameter, beta_1 > 0
    alpha_1, beta_1 = _gamma_params(data=p_data)

    # alpha_2: Distribution 2: shape parameter, alpha_2 > 0
    # beta_2:  Distribution 2: rate or inverse scale parameter, beta_2 > 0  
    alpha_2, beta_2 = _gamma_params(data=q_data)

    # Expression with beta instead of theta
    theta_1 = 1/beta_1
    theta_2 = 1/beta_2
    p1, p2 = 1, 1  # Generalized Gamma Distribution with p=1 is a gamma distribution [1]
    
    # Calculations, see [1] for mathematical details.
    a = p1*(theta_2**alpha_2)*sps.gamma(alpha_2/p2)
    b = p2*(theta_1**alpha_1)*sps.gamma(alpha_1/p1)
    c = (((sps.digamma(alpha_1/p1))/p1) + np.log(theta_1))*(alpha_1 - alpha_2)
    
    # Bi-gamma functions
    d = sps.gamma((alpha_1+p2)/p1)
    e = sps.gamma((alpha_1/p1))
    
    # Calculations
    f = (theta_1/theta_2)**(p2)
    g = alpha_1/p1
    
    # General calculation and output
    r_kld = np.log(a/b) + c + (d/e)*f - g  

    # Final Kullback-Leibler Divergence for Empirically Adjusted Gamma PDFs
    return r_kld

# ----------------------------------------------------------------------------- OHLC FEATURE ENGINEERING -- #
# --------------------------------------------------------------------------------------------------------- #

def ohlc_features(ohlc_data, target_data):
    """
    Feature engineering for OHLC prices. A time check is performed to make sure the targets have exactly the same timestamps as the OHLC data.
    
    
    Parameters
    ----------

    ohlc_data: dict
        Dictionary with the OHLC data for every fold

    target_data: dict
        Dictionary with the target variables for every fold

    """

    # -- For every Fold check form the same timestamp-based index between targets and ohlc
    
    # elements to iterate
    l_target = list(target_data.keys())
    l_ohlc = list(ohlc_data.keys())

    # first validation: The same amount of keys
    if not len(l_target) == len(l_ohlc):
        raise ValueError('The number of elements between targets and ohlc data must be the same')

    # second validation: Exactly the same timestamp-based index for both ohlc and target data
    for i in range(0, len(l_target)):
        if any(list(ohlc_data[i].index) != list(target_data.index[i])):
            raise IndexError('The index in both target and ohlc data must be exactly the same')

    # -- Create linear features
    # -- Create autoregressive features
    # -- Create symbolic features
    # -- Perform information leakage prevention

    return 1

synthetic.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- T-Fold-SV is Time Series Folds for Sequential Validation, the go to alternative for K-Fold-CV       -- #
# -- --------------------------------------------------------------------------------------------------- -- #
# -- Description: Python Implementation of the T-Fold Sequential Validation Method                       -- #
# -- synthetics.py: Synthetic data generation for exploration and testing                                -- #
# -- Author: IFFranciscoME - if.francisco.me@gmail.com                                                   -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- Repository: https://github.com/IFFranciscoME/T-Fold-SV                                              -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

# -- Load libraries for script
import numpy as np

# ------------------------------------------------------------------------------- RANDOM WALK WITH DRIFT -- # 
# ------------------------------------------------------------------------------- ---------------------- -- # 

# Single value
np.random.seed(123) 
mu, sigma = 0.1, 0.1 # mean and standard deviation
random_walk = np.cumsum(np.random.normal(mu, sigma, 100))


continuous_futures.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- T-Fold-SV is Time Series Folds for Sequential Validation, the go to alternative for K-Fold-CV       -- #
# -- --------------------------------------------------------------------------------------------------- -- #
# -- Description: Python Implementation of the T-Fold Sequential Validation Method                       -- #
# -- data.py: read and transform real future contracts data                                              -- #
# -- Author: IFFranciscoME - if.francisco.me@gmail.com                                                   -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- Repository: https://github.com/IFFranciscoME/T-Fold-SV                                              -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""


data.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- T-Fold-SV is Time Series Folds for Sequential Validation, the go to alternative for K-Fold-CV       -- #
# -- --------------------------------------------------------------------------------------------------- -- #
# -- Description: Python Implementation of the T-Fold Sequential Validation Method                       -- #
# -- data.py: read and transform real future contracts data                                              -- #
# -- Author: IFFranciscoME - if.francisco.me@gmail.com                                                   -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- Repository: https://github.com/IFFranciscoME/T-Fold-SV                                              -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

# -- Load packages for this script
import pandas as pd
import numpy as np

# --------------------------------------------------------------------------------- RESAMPLE OHLC PRICES -- #
# ------------------------------------------------------------------------------------------------------ -- #

def resample_data(target_data, target_freq='D', auto_names=True, col_names=None):
    """
    Downsample price data from high frequency, e.g. minute prices, to low frequency, e.g. dayly prices. 

    Parameters
    ----------
    
    target_data: DataFrame

        The OHLC price data to be resampled, currently, only downsample operations are supported, e.g from minute to daily prices, if an upsample operation is detected, e.g. from daily to minute prices, a ValueError is raised. It could be a string to specify the route of the file to be read, or, an existing pd.DataFrame object.

        DataFrame with a column that can be converted to be a timestamp by the .to_datetime() method, if not present a ValueError will be raised. Also other columns are needed, the open, high, low, close prices, and a volume column is optional, all values, except timestamp, will be coherced to numeric. The column names for the output object will stay the same as input object, unless auto_names is set to False and a list of names is provided through col_names parameter.

    target_freq: str

        The target frequency to resample the prices, the options are:
            'H1': to an hour interval, 'H8': to 8 hours interval, 'D': to a day interval.        

    auto_names: bool

        True (Default): Use provided generic columns names: ['timestamp', 'open', 'high', 'low', 'close']
        False: Use the ones in the input object
    
    col_names: list

        None: (Default) 
        list: If auto_names is set to False, a list of every column is expected.

    Returns
    -------    

    r_grouped_data: DataFrame
        
        DataFrame with the new resampled data

    Example
    -------

    >> data = pd.DataFrame({'timestamp': ['2020-01-01 10:00:00', '2020-01-01 10:00:00'],
                            'open': [1.25, 1.25], 'high': [1.35, 1.35],
                            'low': [1.15, 1.15], 'close': [1.30, 1.30], 'vol': [123, 123]})
    >> auto_names = True
    >> col_names = None
    >> target_freq = 'D'

    """
      
    # -- Read input object
    data = target_data.copy()
    
    # -- Column names and types cohercion
    
    # OHLC columns
    if len(list(data.columns)) < 5 or len(list(data.columns)) > 6:
        raise IndexError('There are less than 5 or more than 6 columns in the DataFrame')
    
    # Open, High, Low, Close Columns
    elif len(list(data.columns)) == 5:
        if auto_names:
            data.columns = ['timestamp', 'open', 'high', 'low', 'close']
    
    # Open, High, Low, Close, Volume Columns
    elif len(list(data.columns)) == 6:
        if auto_names:
            data.columns = ['timestamp', 'open', 'high', 'low', 'close', 'volume']
    
    # Timestamp data conversion
    data['timestamp'] = pd.to_datetime(data['timestamp'])
    data.index = data['timestamp']
    data.drop('timestamp', inplace=True, axis=1)

    # -- Resampling process
    conversion = {'open': 'first', 'high': 'max', 'low': 'min', 'close': 'last', 'volume': 'sum'}
    r_grouped_data = data.resample(target_freq).agg(conversion)

    # Eliminar los NAs originados por que ese minuto
    r_grouped_data = r_grouped_data.dropna()

    return r_grouped_data

# ------------------------------------------------------------------------------------ T-FOLDS FORMATION -- #
# --------------------------------------------------------------------------------------------------------- #

def folds_formation(global_data, fold_size):
    """
    Function to separate in T-Folds the data, the functions guarantees not having filtrations
    
    Parameters
    ----------
    global_data : pd.DataFrame
        DataFrame with a global dataset
    
    fold_size : str
        'month': every T-Fold will be of one month of historical data
        'quarter': every T-Fold will be of three months (quarter) of historical data
        'year': every T-Fold will be of twelve months of historical data
        'bi-year': every T-Fold will be of 2 years of historical data
        '80-20': Hold out method, 80% for training and 20% for valing
    
    Returns
    -------
    n_data: dict
        A dictionary with the data separated by Fold, stored by keynames according to Fold size
    
    """

    # For quarterly separation of the data
    if fold_size == 'quarter':
        # List of quarters in the dataset
        quarters = list(set(time.quarter for time in list(global_data.index)))
        # List of years in the dataset
        years = set(time.year for time in list(global_data.index))
        q_data = {}
        # New key for every quarter_year
        for y in sorted(list(years)):
            q_data.update({'q_' + str('0') + str(i) + '_' + str(y) if i <= 9 else str(i) + '_' + str(y):
                               global_data[(global_data.index.year == y) & (global_data.index.quarter == i)]
                           for i in quarters})
        return q_data

    # For semester separation of the data
    elif fold_size == 'semester':
        # List of years in the dataset
        years = set(time.year for time in list(global_data.index))
        s_data = {}
        # New key for every semester_year
        for y in sorted(list(years)):
            s_data.update({'s_' + str('0') + str(1) + '_' + str(y):
                               global_data[(global_data.index.year == y) & ((global_data.index.quarter == 1) |
                                      (global_data.index.quarter == 2))]})

            s_data.update({'s_' + str('0') + str(2) + '_' + str(y):
                               global_data[(global_data.index.year == y) &
                                      ((global_data.index.quarter == 3) | (global_data.index.quarter == 4))]})

        return s_data

    # For yearly separation of the data
    elif fold_size == 'year':
        # List of years in the dataset
        years = set(time.year for time in list(global_data.index))
        y_data = {}
        # New key for every year
        for y in sorted(list(years)):
            y_data.update({'y_' + str(y):
                               global_data[(global_data.index.year == y)]})
                               
        return y_data

    # For yearly separation of the data
    elif fold_size == '80-20':
        # List of years in the dataset
        years = sorted(list(set(time.year for time in list(global_data.index))))
        # dict to store data
        a_8 = int(len(years)*0.80) 
        a_2 = int(len(years)*0.20)
        # data construction
        data_80_20 = {'h_8': global_data[global_data.index.year.isin(years[0:a_8])],
                      'h_2': global_data[global_data.index.year.isin(years[a_8:a_8+a_2])]}
    
        return data_80_20
